File: a004_Median_of_Two_Sorted_Arrays/golang/main.go
package main

import (
	"fmt"
)

func main() {
	a := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	b := []int{10, 20, 30, 40, 50, 60, 70, 80, 90, 100}
	r := findMedianSortedArrays(a, b)
	fmt.Println(r) // 10
}

func findMedianSortedArrays(a, b []int) float64 {
	aLen := len(a)
	bLen := len(b)
	n := aLen + bLen
	all := make([]int, n)
	pos := 0
	i := 0
	j := 0
	for i < aLen && j < bLen {
		if a[i] <= b[j] {
			all[pos] = a[i]
			i++
		} else {
			all[pos] = b[j]
			j++
		}
		pos++
	}
	for i < aLen {
		all[pos] = a[i]
		pos++
		i++
	}
	for j < bLen {
		all[pos] = b[j]
		pos++
		j++
	}
	if n%2 == 1 {
		return float64(all[n/2])
	}
	return float64((all[n/2] + all[n/2-1]) / 2.0)
}

File: a004_Median_of_Two_Sorted_Arrays/rs/src/main.rs
fn main() {
    let a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    let b = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100];
    let r = find_median_sorted_arrays(&a, &b);
    println!("{}", r); // 10
}

fn find_median_sorted_arrays(a: &[i32], b: &[i32]) -> f64 {
    let a_len = a.len();
    let b_len = b.len();
    let n = a_len + b_len;
    let mut all = vec![0; n];
    let mut pos = 0;
    let mut i = 0;
    let mut j = 0;
    while i < a_len && j < b_len {
        if a[i] <= b[j] {
            all[pos] = a[i];
            i += 1;
        } else {
            all[pos] = b[j];
            j += 1;
        }
        pos += 1;
    }
    while i < a_len {
        all[pos] = a[i];
        pos += 1;
        i += 1;
    }
    while j < b_len {
        all[pos] = b[j];
        pos += 1;
        j += 1;
    }
    if n % 2 == 1 {
        return all[n / 2] as f64;
    }
    (all[n / 2] + all[n / 2 - 1]) as f64 / 2.0
}
